src/compiler/checker.ts

```diff
@@ -7662,6 +7662,21 @@ namespace ts {
                 getInitialTypeOfBindingElement(<BindingElement>node);
         }
 
+        function getReferenceFromExpression(node: Expression): Expression {
+            switch (node.kind) {
+                case SyntaxKind.ParenthesizedExpression:
+                    return getReferenceFromExpression((<ParenthesizedExpression>node).expression);
+                case SyntaxKind.BinaryExpression:
+                    switch ((<BinaryExpression>node).operatorToken.kind) {
+                        case SyntaxKind.EqualsToken:

```

Comment:
Nothing other than odd cases like this:

```ts
var x: number | string = getNumberOrString();
x += "hello";
x; // currently has type 'string | number'
```

whereas in this case

```ts
var x: number | string = getNumberOrString();
x = x + "hello";
x; // currently has type 'string'
```

I don't know how often something like that comes up in practice though.